vllm/adapter_commons/models.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, Hashable, Optional, TypeVar
+
+from torch import nn
+
+from vllm.logger import init_logger
+from vllm.utils import LRUCache
+
+logger = init_logger(__name__)
+
+
+class AdapterModel(ABC):
+
+    def __init__(self, model_id=None):
+        self.id = model_id
+
+    @abstractmethod
+    def from_local_checkpoint(cls, model_dir, model_id=None, **kwargs):
+        # Common initialization code
+        # Load weights or embeddings from local checkpoint
+        raise NotImplementedError("Subclasses must implement this method.")
+
+
+T = TypeVar('T')
+
+
+class AdapterLRUCache(LRUCache[T]):
+
+    def __init__(self, capacity: int, deactivate_fn: Callable[[Hashable],
+                                                              None]):
+        super().__init__(capacity)
+        self.deactivate_fn = deactivate_fn
+
+    def _on_remove(self, key: Hashable, value: T):
+        logger.debug("Removing adapter int id: %d", key)
+        self.deactivate_fn(key)
+        return super()._on_remove(key, value)
+
+
+class AdapterModelManager(ABC):
+
+    def __init__(
+        self,
+        model: nn.Module,
+    ):
+        """Create a AdapterModelManager and adapter for a given model.
+        Args:
+            model: the model to be adapted.
+        """
+        self.model: nn.Module = model
+        self._registered_adapters: Dict[int, Any] = {}
+        # Dict instead of a Set for compatibility with LRUCache.
+        self._active_adapters: Dict[int, None] = {}
+        self.adapter_type = 'Adapter'
+        self._last_mapping = None
+
+    def __len__(self) -> int:
+        return len(self._registered_adapters)
+
+    @property
+    @abstractmethod
+    def adapter_slots(self):
+        ...
+
+    @property
+    @abstractmethod
+    def capacity(self):
+        ...
+
+    @abstractmethod
+    def _deactivate_adapter(self, adapter_id: int):
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def deactivate_adapter(self, adapter_id: int) -> bool:
+        if adapter_id in self._active_adapters:
+            self._deactivate_adapter(adapter_id)
+            self._active_adapters.pop(adapter_id)
+            return True
+        return False
+
```

Comment:
let's just make the public APIs abstract and push the implementation to subclasses

Reply:
The code here is common between both lora and prompt adapter, do we still reimplement this in both subclasses?

Reply:
Hey @Yard1 the code here is common between both lora and prompt adapter, do we still reimplement this in both subclasses?

Reply:
let's use util functions for code reuse

Reply:
these functions mainly call other functions and internal objects, so won't be able to reuse some of them

Reply:
actually scratch that, we pass the references to the class objects and they should be updated as well